### DISCUSSION TITLE
Which layer is this designed to work with/in?

### DISCUSSION THREAD

**Discussion by mpenna** [dxdc/aws-sdk-google-apps#16] [Q&A]
Hi. I haven't tried the repo yet, so I'm a bit confused regarding the actual execution context of this library. From the various "library" mentions throughout the README, along with some mentions to be found in the issues section, I guess this SDK is supposed to be consumed at the "server" side, right? The confusion comes from a couple other mentions regarding the "browser context". Could you clarify this a bit?

BTW, congratulations on this awesome work: tackling the AWS SDK is indeed one huge feat!

**Comment by dxdc** [comment]
This runs with [Google Apps Script](https://www.google.com/script/start/).

AWS-SDK was built for browsers that use the standard Web API's, yet Google Apps Script lacks access to many of those (see: https://developer.mozilla.org/en-US/docs/Web/API). This library is essentially a polyfill for some of those, rather than trying to re-create the wheel and build various features of AWS-SDK from scratch.

**Comment by mpenna** [comment]
Thanks for the quick reply. Let me see if I got this right. 

The original AWS SDK for Javascript is designed to work in a Nodejs context/environment, right? This repo, on the other hand, is a polyfill that opens up the possibility of using those same APIs but in a GAS context, right? However, my original question has to do with the presumption that the GAS context is actually twofold: 

- the browser context (e.g.: a Javascript frontend/web project deployed as a sheet's sidebar or modal dialog, or perhaps even as a web app)
- the server context (e.g.: scripts that run at the "backend/server" and that make use of GAS's own range of APIs - usually invoked from the frontend via the `google.script.run` API)

Looking around for GAS boilerplates and starter projects, one finds various frontend web projects that make use of both contexts (i.e.: frontend/browser code calling GAS/backend APIs).

So, given all this, is it safe to assume that this pollyfil's execution context is that of GAS's backend/server context?

**Reply by dxdc** [reply]
> So, given all this, is it safe to assume that this pollyfil's execution context is that of GAS's backend/server context?

Yes. I think you should familiarize yourself with how GAS works. It's "browser" context is highly limited (sidebar/modal dialog).

A patched version of the AWS SDK isn't required for Node or web modules.

**Comment by mpenna** [comment]
Just to illustrate, this one here is a repo that puts this concept of frontend/backend for GAS projects into practice. The `frontend` module/package is a conventional Vue.js web app slightly adapted to work under and around the limitations of a Google document's browser context (e.g.: sidebar). On the other hand, the `backend` is the one that holds the GAS server-side logic, where all of their native APIs will be available.

https://github.com/fossamagna/gas-vue-boilerplate

I'm using this as a starting point for my project where some of its features will rely on being able to interact with a couple of AWS's APIs. So my previous questions were all framed considering this scenario and overall project structure/architecture. Then I guess that this SDK of yours would have to be integrated with what this other repo refers to as "the backend", right? 

Once again, thanks. And sorry for all these questions.

**Reply by dxdc** [reply]
> Then I guess that this SDK of yours would have to be integrated with what this other repo refers to as "the backend", right?

Yes. Their frontend is highly limited. See Google's documentation: https://developers.google.com/apps-script/guides/html/communication#code.gs for how the communication between the client/server works.
